app/src/main/java/com/poklad/androidifystore/presentation/ui/screens/all_products/AllProductsFragment.kt

```diff
@@ -0,0 +1,103 @@
+package com.poklad.androidifystore.presentation.ui.screens.all_products
+
+import android.content.Context
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.widget.Toast
+import androidx.fragment.app.viewModels
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.repeatOnLifecycle
+import androidx.navigation.fragment.findNavController
+import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.poklad.androidifystore.R
+import com.poklad.androidifystore.StoreApp
+import com.poklad.androidifystore.databinding.FragmentAllProductsBinding
+import com.poklad.androidifystore.domain.model.ProductItem
+import com.poklad.androidifystore.presentation.ui.base.BaseFragment
+import com.poklad.androidifystore.presentation.ui.base.BaseViewModel
+import com.poklad.androidifystore.utils.Resource
+import com.poklad.androidifystore.utils.invisible
+import com.poklad.androidifystore.utils.log
+import com.poklad.androidifystore.utils.visible
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+class AllProductsFragment : BaseFragment<FragmentAllProductsBinding, BaseViewModel>() {
+
+    @Inject
+    lateinit var viewModelFactory: ViewModelProvider.Factory
+
+    override val viewModel: AllProductsViewModel by viewModels {
+        viewModelFactory
+    }
+    private val allProductsAdapter: AllProductsAdapter by lazy {
+        AllProductsAdapter()
+    }
+
+    override fun inflateViewBinding(inflater: LayoutInflater): FragmentAllProductsBinding =
+        FragmentAllProductsBinding.inflate(inflater)
+
+    override fun onAttach(context: Context) {
+        super.onAttach(context)
+        StoreApp.daggerComponent.inject(this)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        initRecyclerView()
+        setUpObserver()
+    }
+
+    private fun setUpObserver() {
+        viewLifecycleOwner.lifecycleScope.launch {
+            repeatOnLifecycle(Lifecycle.State.STARTED) {
+                viewModel.products.collect { resource ->
+                    when (resource) {
+                        is Resource.Success -> {
```

Comment:
handle the logic in `viewModel`, pass into your view handled data. in that case you need to create f.e. three flow objects loadingFlow, productFlow and errorFlow. loadingFlow and errorFlow you can move into `BaseViewModel`

Reply:
I saw somewhere, I don't remember where, but I think it was a strong developer who said that all logic should be moved to UseCase. And just catch the result in the viewmodel. And flow we can use in domain becouce this is chapter of Kotlin